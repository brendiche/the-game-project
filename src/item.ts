import { Engine } from "./gameEngine";
import { getPosition, setPosition } from "./helper";

const ITEM_CONFIG = {
  movingStep: 10,
  maxDistance: 500,
}

export const createItem = (engine: Engine, className: string, style:Partial<CSSStyleDeclaration>) => {
  const element = document.createElement('div');
  element.className = className;
  for(const prop in style){
    element.style[prop] = style[prop];
  }
  const initialPosition = getPosition(element);
  engine.addGamingThread(() => {
    if(getPosition(element) >= initialPosition + ITEM_CONFIG.maxDistance){
      element.remove();
    }else{
      setPosition(element, getPosition(element) + ITEM_CONFIG.movingStep);
    }
  });
  document.body.appendChild(element)
  return element;
}