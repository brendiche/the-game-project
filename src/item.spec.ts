/**
 * @jest-environment jsdom
 */
 const dateNowStub = () => 1644182219303;
 global.Date.now = dateNowStub;
import { Item } from "./item"
const style = {
  position: 'absolute',
  left: '10px',
  top: '10px',
} as CSSStyleDeclaration

const item = new Item({className:'test',side:'left', style:style});
describe('item', () => {
  it('should create an item', () => {
    expect(item).toMatchSnapshot();
    expect(item.element).toMatchSnapshot();
    expect(item.id).toMatchSnapshot();
    expect(item.position).toMatchSnapshot();
    expect(item.initialPosition).toMatchSnapshot();
  });
  it('should remove element', ()=> {
    item.remove()
    expect(item.element).toMatchSnapshot()
  })
})