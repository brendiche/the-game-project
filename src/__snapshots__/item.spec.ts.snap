// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item should create an item 1`] = `
Item {
  "_element": <div
    class="test"
    style="position: absolute; left: 10px; top: 10px;"
  />,
  "_id": 1644182219303,
  "_initialPosition": 10,
}
`;

exports[`item should create an item 2`] = `
<div
  class="test"
  style="position: absolute; left: 10px; top: 10px;"
/>
`;

exports[`item should create an item 3`] = `1644182219303`;

exports[`item should create an item 4`] = `10`;

exports[`item should create an item 5`] = `10`;

exports[`item should remove element 1`] = `
<div
  class="test"
  style="position: absolute; left: 10px; top: 10px;"
/>
`;
